core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala

```diff
@@ -179,7 +180,11 @@ class HistoryServer(
   }
```

Comment:
So, should `getApplicationList()` even exist? it sounds like you'd never really want to get the whole materialized list of apps. It looks like it's either called and filtered, or called for its iterator. Is a deeper improvement here to just return an iterator from the provider?

Looking deeper, I suppose one issue here is that the underlying data structure this comes from is mutable and changing. But then I think that means any view-based or iterator-based result runs the risk of traversing it while it's modified. Hm.

Maybe it's still no worse to do so, to return an iterator and immediately filter to get the right subset copy, and proceed?